File: python/euler/bin/p0059.py
# project euler: problem 59

import operator
import re
from collections.abc import Collection, Iterable
from itertools import product
from typing import TextIO


def decode(cipher_text: Collection[int], key: tuple[int, ...]) -> list[int]:
    text_len = len(cipher_text)
    key_len = len(key)
    key_lst = (key * (text_len // key_len + 1))[:text_len]

    return list(map(operator.xor, cipher_text, key_lst))


def calc_score(lst: Iterable[int]) -> int:
    word_list = [s.casefold() for s in re.split(r'[,. ]', ''.join(map(chr, lst)))]
    cnt = 0
    for word in ['and', 'of', 'a', 'to', 'in']:
        cnt += word_list.count(word)

    return cnt


def compute(fh: TextIO) -> str:
    cipher_text = list(map(int, fh.read().split(',')))
    key_lst = product(range(ord('a'), ord('z') + 1), repeat=3)

    max_score: tuple[int, tuple[int, ...], list[int]] = (0, (), [])
    for key in key_lst:
        plain_text = decode(cipher_text, key)
        if (score := calc_score(plain_text)) > max_score[0]:
            max_score = (score, key, plain_text)

    return str(sum(max_score[2]))


def solve() -> str:
    from euler.lib.resource import asset_file

    fh = asset_file('p059_cipher.txt')
    result = compute(fh)
    fh.close()
    return result

File: typescript/euler/bin/p0059.ts
// project euler: problem 59

import { permutationsWithReplacement } from "combinatorics/mod.ts";
import { assetData } from "../lib/asset.ts";
import { sum } from "../lib/math.ts";
import { range } from "../lib/util.ts";

const decode = (cipherText: [number, number][], key: number[]): number[] =>
  cipherText.map(([e, i]) => e ^ key[i]);

const calcScore = (lst: number[]): number => {
  const wordLst = lst.map((x) => String.fromCharCode(x)).join("").split(/[,. ]/)
    .map((x) => x.toLocaleLowerCase());

  let cnt = 0;
  for (const word of ["and", "or", "a", "to", "in"]) {
    cnt += wordLst.filter((x) => x === word).length;
  }

  return cnt;
};

export const compute = (data: string): string => {
  const cipherText: [number, number][] = data.split(",").map((
    x,
    idx,
  ) => [Number(x), idx % 3]);
  let maxScore: [number, number[], number[]] = [0, [], []];

  for (
    const key of permutationsWithReplacement(
      range("a".charCodeAt(0), "z".charCodeAt(0) + 1),
      3,
    )
  ) {
    const plainText = decode(cipherText, key);
    const score = calcScore(plainText);
    if (score > maxScore[0]) {
      maxScore = [score, key, plainText];
    }
  }
  return String(sum(maxScore[2]));
};

export const solve = (): string => {
  const data = new TextDecoder().decode(assetData("p059_cipher.txt"));
  return compute(data);
};
